StreamChatClient_v3/WebSocketClient/WebSocketClient.swift

```diff
@@ -209,18 +206,32 @@ protocol ConnectionStateDelegate: AnyObject {
 extension WebSocketClient {
     /// An object encapsulating all dependencies of `WebSocketClient`.
     struct Environment {
-        var timer: Timer.Type = DefaultTimer.self
+        typealias CreatePingController = (
+            _ timerType: Timer.Type,
+            _ timerQueue: DispatchQueue,
+            _ ping: @escaping WebSocketPingController.PingAction,
+            _ forceReconnect: @escaping WebSocketPingController.ForceReconnectAction
+        ) -> WebSocketPingController
         
-        var notificationCenterBuilder: () -> NotificationCenter = NotificationCenter.init
+        typealias CreateEngine = (
+            _ request: URLRequest,
+            _ sessionConfiguration: URLSessionConfiguration,
+            _ callbackQueue: DispatchQueue
+        ) -> WebSocketEngine
         
-        var createEngine: (_ request: URLRequest, _ sessionConfiguration: URLSessionConfiguration, _ callbackQueue: DispatchQueue)
-            -> WebSocketEngine = {
-                if #available(iOS 13, *) {
-                    return URLSessionWebSocketEngine(request: $0, sessionConfiguration: $1, callbackQueue: $2)
-                } else {
-                    return StarscreamWebSocketProvider(request: $0, sessionConfiguration: $1, callbackQueue: $2)
-                }
+        var timerType: Timer.Type = DefaultTimer.self
+        
+        var createPingController: CreatePingController = WebSocketPingController.init
+        
+        var createNotificationCenter: () -> NotificationCenter = NotificationCenter.init
+        
+        var createEngine: CreateEngine = {
+            if #available(iOS 13, *) {
+                return URLSessionWebSocketEngine(request: $0, sessionConfiguration: $1, callbackQueue: $2)
+            } else {
+                return StarscreamWebSocketProvider(request: $0, sessionConfiguration: $1, callbackQueue: $2)
             }
+        }
```

Comment:
You should rebase, this part has changes IIRC

Reply:
What exactly changed?